### PR TITLE
[WIP] Syntax light

### PR DESCRIPTION
**Pull request by Joeper214**
### Syntax Highlighting on CF_templates JSON.

### CODE REVIEW

**Comment by Joeper214** [comment]
test is taking too long to finish.

**Comment by pocke** [comment]
Yes, I think coveralls doesn't update check status.

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -26,3 +26,6 @@
 //= require utils
 //= require server_status
 //= require notification_center
+//= require ace-rails-ap
+//= require ace/mode-javascript
+//= require ace/mode-json
```
Comment (pocke) [inline app/assets/javascripts/application.js]: SkyHopper load third party library of JS from `app/assets/javascripts/libraries.js`.
Please move those requires from application.js to libraries.js

Comment (Joeper214) [inline app/assets/javascripts/application.js]: :+1:

File: app/assets/javascripts/infrastructures.js [hunk 2]
```diff
@@ -123,19 +125,40 @@
         self.result.name   = cft.name;
         self.result.detail = cft.detail;
         self.result.value  = cft.value;
+        this.editor(cft.value);
```
Comment (pocke) [inline app/assets/javascripts/infrastructures.js]: I think `this.editor()` doesn't need any arguments.

Comment (Joeper214) [inline app/assets/javascripts/infrastructures.js]: Yes. I forgot. thanks.

File: app/assets/javascripts/infrastructures.js [hunk 3]
```diff
@@ -123,19 +125,40 @@
...
+      }
     },
     computed: {
       jsonParseErr: function () { return jsonParseErr(this.result.value); },
     },
     created: function () {
       console.log(this);
-    }
+      this.loading = true;
+      setTimeout(this.editor, 1000);
```
Comment (pocke) [inline app/assets/javascripts/infrastructures.js]: Why do you use `setTimeout`?

Comment (pocke) [inline app/assets/javascripts/infrastructures.js]: Oh, try to remove setTimeout, doesnt work...

Comment (Joeper214) [inline app/assets/javascripts/infrastructures.js]: Because the div takes too long to load. it causes the ace editor to have this error. 

ace_errr

Comment (pocke) [inline app/assets/javascripts/infrastructures.js]:
```diff
diff --git a/app/assets/javascripts/infrastructures.js b/app/assets/javascripts/infrastructures.js
index 3f0f017..f2ddcd0 100644
--- a/app/assets/javascripts/infrastructures.js
+++ b/app/assets/javascripts/infrastructures.js
@@ -157,8 +157,10 @@
     created: function () {
       console.log(this);
       this.loading = true;
-      setTimeout(this.editor, 1000);
     },
+    ready: function () {
+      this.editor();
+    }
   });
 
   Vue.component("insert-cf-params", {

```

when `created`, there isn't `#description` in DOM yet. so, if remove `settimeout`, error is raised.

We can use the `ready` function. When `ready`, there is `#description` in DOM already.
http://vuejs.org/api/options.html#ready

Prease replace `created` to `ready`.
And loading is not necessary?

Comment (Joeper214) [inline app/assets/javascripts/infrastructures.js]: alright. Thanks. I think ready is able to listen if all the tabpane elements are loaded. just like jQuery $.ready function. I'll change it right away.

**Comment by Joeper214** [comment]
I think we also can use this feature to Serverspec Generator right?

**Comment by pocke** [comment]
I think we should separate ace editor the Ace from the add_modify component.
And separate module file.

for example

```javascript
Vue.compoent('ace-editor', {
props: {
value: ... // => two way binding
mode: ... // => json, ruby, javascript, ...
},
...
})
```

```html
<ace-editor value="{{@ result.value}}" mode="json"></ace-editor>
```

> I think we also can use this feature to Serverspec Generator right?

Yes!


http://vuejs.org/guide/components.html#Prop_Binding_Types

**Comment by Joeper214** [comment]
hmmm. ..  but ace editor is using the div binding with an ID on it. can this be achieved using a vue component? 

ex: 

`var editor = ace.edit("description");`

"description is a div element" who to bind this with <ace-editor> component?

**Comment by pocke** [comment]
Ace editor can use HTMLElement as parameter of edit method.
http://ace.c9.io/#nav=api&api=ace

And Vue.js can get HTMLElement by `$el` property. http://vuejs.org/api/instance-properties.html#vm-\$el

```javascript
// in component function
var editor = ace.edit(this.$el)
```

**Comment by Joeper214** [comment]
I think there's a problem when selecting the cf_templates from the list of presets. because the data is not updating on the ace-editor component.  Is there a way that I can update or trigger the data in ace-editor using the parent node?

**Comment by Joeper214** [comment]
I tried using this code but it seems lacking

###Infra.js
`Vue.directive("ace-editor", {
    props: {
      value : null,
      twoWayProp: {
        twoWay: true
      },
    },
    template: '<div id="description" style="position: absolute; width: 100%; left: 0px; height: 100%; font-size: 12px;"></div>',
  
    replace: true,
    ready: function () {
      this.editor();
    },
    methods: {
      editor: function () {
        var self = this;
        var editor = ace.edit('description');
        editor.setTheme("ace/theme/github");
        editor.getSession().setMode("ace/mode/json");
        editor.getSession().setUseWrapMode(true);
        console.log(self.$data.value);
        if (self.$data.value) {
          editor.getSession().setValue(self.$data.value);
        }else {
          editor.getSession().setValue();
        }
        editor.getSession().on('change', function(){
          self.$data.value = editor.getSession().getValue();
        });
      },
    },
  });`


###add_modify.html
```<ace-editor value={{@result.value}}></ace-editor>```

**Comment by Joeper214** [comment]
I think I got it!..

With this code, I can reuse ace as a Directive.

`Vue.directive("ace", {
      twoWay: true,
      bind: function () {
          // 请自行初始化CodeMirror, 可使用 $(this.el).data('codemirror')
          console.log(this.el);
          this.editor = ace.edit(this.el);
          this.editor.setTheme("ace/theme/github");
          this.editor.getSession().setMode("ace/mode/json");
          this.editor.getSession().setUseWrapMode(true);
          // model主动赋值标记
          this.silent = false;
          this.handler = function () {
              if (!this.silent) {
                  this.set(this.editor.getSession().getValue(), true); // 加锁，防止相互赋值
              }
          }.bind(this);
          this.editor.on("change", this.handler);
      },
      update: function (value, oldValue) {
          this.silent = true;
          this.editor.getSession().setValue(value);
          this.silent = false;
      }
  });`

**Comment by pocke** [comment]
LGTM :+1:

**Comment by Joeper214** [comment]
Thanks!

**Pull request merged**